Different treatment for `?:` and `if...else` in dataflow refinement

Consider the following code.  The two classes `ImmutableIntList1` and `ImmutableIntList2` have the same logic and control flow, but the Nullness Checker issues an error message for the first but not the second:

```
Issue3622.java:14: error: [contracts.conditional.postcondition.not.satisfied] conditional postcondition is not satisfied when equals returns true.
            return obj instanceof ImmutableIntList1 ? true : obj instanceof List;
            ^
  found   : obj is @Nullable
  required: obj is @NonNull
```

```java
// Test case for https://tinyurl.com/cfissue/3622

// @skip-test until the issue is fixed

import java.util.List;
import org.checkerframework.checker.nullness.qual.Nullable;

public class Issue3622 {

    public class ImmutableIntList1 {

        @Override
        public boolean equals(@Nullable Object obj) {
            return obj instanceof ImmutableIntList1 ? true : obj instanceof List;
        }
    }

    public class ImmutableIntList2 {

        @Override
        public boolean equals(@Nullable Object obj) {
            if (obj instanceof ImmutableIntList2) {
                return true;
            } else {
                return obj instanceof List;
            }
        }
    }
}
```

--- a/checker/tests/nullness/Issue3622.java
+++ b/checker/tests/nullness/Issue3622.java
@@ -7,23 +7,107 @@ import org.checkerframework.checker.nullness.qual.Nullable;
 
 public class Issue3622 {
 
+    // These currently pass (no warnings)
+
     public class ImmutableIntList1 {
 
         @Override
-        public boolean equals(@Nullable Object obj) {
-            return obj instanceof ImmutableIntList1 ? true : obj instanceof List;
+        public boolean equals(@Nullable Object obj1) {
+            if (obj1 instanceof ImmutableIntList1) {
+                return true;
+            } else {
+                return obj1 instanceof List;
+            }
         }
     }
 
     public class ImmutableIntList2 {
 
         @Override
-        public boolean equals(@Nullable Object obj) {
-            if (obj instanceof ImmutableIntList2) {
+        public boolean equals(@Nullable Object obj2) {
+            return obj2 instanceof ImmutableIntList2;
+        }
+    }
+
+    public class ImmutableIntList3 {
+
+        @Override
+        public boolean equals(@Nullable Object obj3) {
+            if (obj3 instanceof ImmutableIntList3) {
                 return true;
             } else {
-                return obj instanceof List;
+                return false;
             }
+        }
+    }
+
+    public class ImmutableIntList7 {
+
+        @Override
+        public boolean equals(@Nullable Object obj7) {
+            return obj7 instanceof ImmutableIntList7 ? true : false;
+        }
+    }
+
+    public class ImmutableIntList8 {
+
+        @Override
+        public boolean equals(@Nullable Object obj8) {
+            return obj8 instanceof ImmutableIntList8 ? true : obj8 instanceof List;
+        }
+    }
+
+    public class ImmutableIntList9 {
+
+        @Override
+        public boolean equals(@Nullable Object obj9) {
+            return obj9 instanceof ImmutableIntList9
+                    ? obj9 instanceof ImmutableIntList9
+                    : obj9 instanceof ImmutableIntList9;
+        }
+    }
+
+    // These currently fail (false positive warnings)
+
+    public class ImmutableIntList4 {
+
+        @Override
+        @SuppressWarnings(
+                "contracts.conditional.postcondition.not.satisfied" // TODO: `if` needs the
+        // BOTH_TO_THEN treatment that ?: gets.
+        )
+        public boolean equals(@Nullable Object obj4) {
+            boolean b;
+            if (obj4 instanceof ImmutableIntList4) {
+                b = true;
+            } else {
+                b = false;
+            }
+            return b;
+        }
+    }
+
+    public class ImmutableIntList5 {
+
+        @Override
+        @SuppressWarnings(
+                "contracts.conditional.postcondition.not.satisfied" // TODO: Need special treatment
+        // for true and false boolean  literals (cut off dead parts of graph).
+        )
+        public boolean equals(@Nullable Object obj5) {
+            return true ? obj5 instanceof ImmutableIntList5 : obj5 instanceof ImmutableIntList5;
+        }
+    }
+
+    public class ImmutableIntList6 {
+
+        @Override
+        @SuppressWarnings(
+                "contracts.conditional.postcondition.not.satisfied" // TODO: Need special treatment
+        // for true and false boolean  literals (cut off dead parts of graph).
+        )
+        public boolean equals(@Nullable Object obj6) {
+            return true ? obj6 instanceof ImmutableIntList6 : false;
         }
     }
 }

--- a/dataflow/src/main/java/org/checkerframework/dataflow/analysis/ForwardAnalysisImpl.java
+++ b/dataflow/src/main/java/org/checkerframework/dataflow/analysis/ForwardAnalysisImpl.java
@@ -447,6 +447,22 @@ public class ForwardAnalysisImpl<
                         Store.Kind.ELSE,
                         addToWorklistAgain);
                 break;
+            case BOTH_TO_THEN:
+                addStoreBefore(
+                        succ,
+                        node,
+                        currentInput.getRegularStore(),
+                        Store.Kind.THEN,
+                        addToWorklistAgain);
+                break;
+            case BOTH_TO_ELSE:
+                addStoreBefore(
+                        succ,
+                        node,
+                        currentInput.getRegularStore(),
+                        Store.Kind.ELSE,
+                        addToWorklistAgain);
+                break;
         }
     }
 

--- a/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Store.java
+++ b/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Store.java
@@ -25,14 +25,23 @@ public interface Store<S extends Store<S>> {
 
     /** A flow rule describes how stores flow along one edge between basic blocks. */
     public static enum FlowRule {
-        EACH_TO_EACH, // The normal case, then store flows to the then store
-        // and else store flows to the else store.
-        THEN_TO_BOTH, // Then store flows to both then and else of successor.
-        ELSE_TO_BOTH, // Else store flows to both then and else of successor.
-        THEN_TO_THEN, // Then store flows to the then of successor.  Else store is ignored.
-        ELSE_TO_ELSE, // Else store flows to the else of successor.  Then store is ignored.
-        BOTH_TO_THEN, // Both stores flow to the then of successor.
-        BOTH_TO_ELSE, // Both stores flow to the else of successor.
+        /**
+         * The normal case: then store flows to the then store, and else store flows to the else
+         * store.
+         */
+        EACH_TO_EACH,
+        /** Then store flows to both then and else of successor. */
+        THEN_TO_BOTH,
+        /** Else store flows to both then and else of successor. */
+        ELSE_TO_BOTH,
+        /** Then store flows to the then of successor. Else store is ignored. */
+        THEN_TO_THEN,
+        /** Else store flows to the else of successor. Then store is ignored. */
+        ELSE_TO_ELSE,
+        /** Both stores flow to the then of successor. */
+        BOTH_TO_THEN,
+        /** Both stores flow to the else of successor. */
+        BOTH_TO_ELSE,
     }
 
     /**

--- a/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Store.java
+++ b/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Store.java
@@ -31,6 +31,8 @@ public interface Store<S extends Store<S>> {
         ELSE_TO_BOTH, // Else store flows to both then and else of successor.
         THEN_TO_THEN, // Then store flows to the then of successor.  Else store is ignored.
         ELSE_TO_ELSE, // Else store flows to the else of successor.  Then store is ignored.
+        BOTH_TO_THEN, // Both stores flow to the then of successor.
+        BOTH_TO_ELSE, // Both stores flow to the else of successor.
     }
 
     /**

--- a/dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java
+++ b/dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java
@@ -95,6 +95,7 @@ import javax.lang.model.type.UnionType;
 import javax.lang.model.util.Elements;
 import javax.lang.model.util.Types;
 import org.checkerframework.checker.interning.qual.FindDistinct;
+import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.dataflow.analysis.Store;
 import org.checkerframework.dataflow.cfg.CFGBuilder.ExtendedNode.ExtendedNodeType;
@@ -544,20 +545,43 @@ public class CFGBuilder {
         /** The jump target label. */
         protected final Label jumpTarget;
 
+        /** The flow rule for this edge. */
+        protected final Store.FlowRule flowRule;
+
         /**
          * Construct an UnconditionalJump.
          *
          * @param jumpTarget the jump target label
          */
         public UnconditionalJump(Label jumpTarget) {
+            this(jumpTarget, Store.FlowRule.EACH_TO_EACH);
+        }
+
+        /**
+         * Construct an UnconditionalJump, specifying its flow rule.
+         *
+         * @param jumpTarget the jump target label
+         * @param flowRule the flow rule for this edge
+         */
+        public UnconditionalJump(Label jumpTarget, Store.FlowRule flowRule) {
             super(ExtendedNodeType.UNCONDITIONAL_JUMP);
             assert jumpTarget != null;
             this.jumpTarget = jumpTarget;
+            this.flowRule = flowRule;
         }
 
         @Override
         public Label getLabel() {
             return jumpTarget;
+        }
+
+        /**
+         * Returns the flow rule for this edge.
+         *
+         * @return the flow rule for this edge
+         */
+        public Store.FlowRule getFlowRule() {
+            return flowRule;
         }
 
         /**
@@ -1249,25 +1273,76 @@ public class CFGBuilder {
     /* Phase Two */
     /* --------------------------------------------------------- */
 
-    /** Tuple class with up to three members. */
-    protected static class Tuple<A, B, C> {
-        public final A a;
-        public final B b;
-        public final C c;
+    /** Represents a missing edge that will be added later. */
+    protected static class MissingEdge {
+        /** The source of the edge. */
+        final SingleSuccessorBlockImpl source;
+        /** The index (target?) of the edge. Null means go to exceptional exit. */
+        final @Nullable Integer index;
+        /** The cause exception type, for an exceptional edge; otherwise null. */
+        final @Nullable TypeMirror cause;
 
-        public Tuple(A a, B b) {
-            this(a, b, null);
+        /** The flow rule for this edge. */
+        final Store.@Nullable FlowRule flowRule;
+
+        /**
+         * Create a new MissingEdge.
+         *
+         * @param source the source of the edge
+         * @param index the index (target?) of the edge
+         */
+        public MissingEdge(SingleSuccessorBlockImpl source, int index) {
+            this(source, index, null, Store.FlowRule.EACH_TO_EACH);
         }
 
-        public Tuple(A a, B b, C c) {
-            this.a = a;
-            this.b = b;
-            this.c = c;
+        /**
+         * Create a new MissingEdge.
+         *
+         * @param source the source of the edge
+         * @param index the index (target?) of the edge
+         * @param flowRule the flow rule for this edge
+         */
+        public MissingEdge(SingleSuccessorBlockImpl source, int index, Store.FlowRule flowRule) {
+            this(source, index, null, flowRule);
+        }
+
+        /**
+         * Create a new MissingEdge.
+         *
+         * @param source the source of the edge
+         * @param index the index (target?) of the edge; null means go to exceptional exit
+         * @param cause the cause exception type, for an exceptional edge; otherwise null
+         */
+        public MissingEdge(
+                SingleSuccessorBlockImpl source,
+                @Nullable Integer index,
+                @Nullable TypeMirror cause) {
+            this(source, index, cause, Store.FlowRule.EACH_TO_EACH);
+        }
+
+        /**
+         * Create a new MissingEdge.
+         *
+         * @param source the source of the edge
+         * @param index the index (target?) of the edge; null means go to exceptional exit
+         * @param cause the cause exception type, for an exceptional edge; otherwise null
+         * @param flowRule the flow rule for this edge
+         */
+        public MissingEdge(
+                SingleSuccessorBlockImpl source,
+                @Nullable Integer index,
+                @Nullable TypeMirror cause,
+                Store.FlowRule flowRule) {
+            assert (index != null) || (cause != null);
+            this.source = source;
+            this.index = index;
+            this.cause = cause;
+            this.flowRule = flowRule;
         }
 
         @Override
         public String toString() {
-            return "Tuple<" + a + ", " + b + ", " + c + ">";
+            return "MissingEdge(" + source + ", " + index + ", " + cause + ")";
         }
     }
 
@@ -1289,6 +1364,7 @@ public class CFGBuilder {
 
             Map<Label, Integer> bindings = in.bindings;
             ArrayList<ExtendedNode> nodeList = in.nodeList;
+            // A leader is an extended node which will give rise to a basic block in phase two.
             Set<Integer> leaders = in.leaders;
 
             assert !in.nodeList.isEmpty();
@@ -1299,20 +1375,17 @@ public class CFGBuilder {
                     new SpecialBlockImpl(SpecialBlockType.EXCEPTIONAL_EXIT);
 
             // record missing edges that will be added later
-            Set<Tuple<? extends SingleSuccessorBlockImpl, Integer, ?>> missingEdges =
-                    new MostlySingleton<>();
+            Set<MissingEdge> missingEdges = new MostlySingleton<>();
 
             // missing exceptional edges
-            Set<Tuple<ExceptionBlockImpl, Integer, TypeMirror>> missingExceptionalEdges =
-                    new LinkedHashSet<>();
+            Set<MissingEdge> missingExceptionalEdges = new LinkedHashSet<>();
 
             // create start block
             SpecialBlockImpl startBlock = new SpecialBlockImpl(SpecialBlockType.ENTRY);
-            missingEdges.add(new Tuple<>(startBlock, 0));
+            missingEdges.add(new MissingEdge(startBlock, 0));
 
-            // loop through all 'leaders' (while dynamically detecting the
-            // leaders)
-            RegularBlockImpl block = new RegularBlockImpl();
+            // Loop through all 'leaders' (while dynamically detecting the leaders).
+            @NonNull RegularBlockImpl block = new RegularBlockImpl(); // block being processed/built
             int i = 0;
             for (ExtendedNode node : nodeList) {
                 switch (node.getType()) {
@@ -1357,7 +1430,7 @@ public class CFGBuilder {
                             Integer target = bindings.get(thenLabel);
                             assert target != null;
                             missingEdges.add(
-                                    new Tuple<>(
+                                    new MissingEdge(
                                             new RegularBlockImpl() {
                                                 @Override
                                                 public void setSuccessor(BlockImpl successor) {
@@ -1368,7 +1441,7 @@ public class CFGBuilder {
                             target = bindings.get(elseLabel);
                             assert target != null;
                             missingEdges.add(
-                                    new Tuple<>(
+                                    new MissingEdge(
                                             new RegularBlockImpl() {
                                                 @Override
                                                 public void setSuccessor(BlockImpl successor) {
@@ -1379,6 +1452,7 @@ public class CFGBuilder {
                             break;
                         }
                     case UNCONDITIONAL_JUMP:
+                        UnconditionalJump uj = (UnconditionalJump) node;
                         if (leaders.contains(i)) {
                             RegularBlockImpl b = new RegularBlockImpl();
                             block.setSuccessor(b);
@@ -1387,12 +1461,13 @@ public class CFGBuilder {
                         node.setBlock(block);
                         if (node.getLabel() == in.regularExitLabel) {
                             block.setSuccessor(regularExitBlock);
+                            block.setFlowRule(uj.getFlowRule());
                         } else if (node.getLabel() == in.exceptionalExitLabel) {
                             block.setSuccessor(exceptionalExitBlock);
+                            block.setFlowRule(uj.getFlowRule());
                         } else {
-                            Integer target = bindings.get(node.getLabel());
-                            assert target != null;
-                            missingEdges.add(new Tuple<>(block, target));
+                            int target = bindings.get(node.getLabel());
+                            missingEdges.add(new MissingEdge(block, target, uj.getFlowRule()));
                         }
                         block = new RegularBlockImpl();
                         break;
@@ -1410,7 +1485,7 @@ public class CFGBuilder {
                         // Note: do not link to the next block for throw statements
                         // (these throw exceptions for sure)
                         if (!node.getTerminatesExecution()) {
-                            missingEdges.add(new Tuple<>(e, i + 1));
+                            missingEdges.add(new MissingEdge(e, i + 1));
                         }
 
                         // exceptional edges
@@ -1421,7 +1496,7 @@ public class CFGBuilder {
                                 Integer target = bindings.get(label);
                                 // TODO: This is sometimes null; is this a problem?
                                 // assert target != null;
-                                missingExceptionalEdges.add(new Tuple<>(e, target, cause));
+                                missingExceptionalEdges.add(new MissingEdge(e, target, cause));
                             }
                         }
                         break;
@@ -1430,20 +1505,23 @@ public class CFGBuilder {
             }
 
             // add missing edges
-            for (Tuple<? extends SingleSuccessorBlockImpl, Integer, ?> p : missingEdges) {
-                Integer index = p.b;
-                assert index != null : "CFGBuilder: problem in CFG construction " + p.a;
+            for (MissingEdge p : missingEdges) {
+                Integer index = p.index;
+                assert index != null : "CFGBuilder: problem in CFG construction " + p.source;
                 ExtendedNode extendedNode = nodeList.get(index);
                 BlockImpl target = extendedNode.getBlock();
-                SingleSuccessorBlockImpl source = p.a;
+                SingleSuccessorBlockImpl source = p.source;
                 source.setSuccessor(target);
+                if (p.flowRule != null) {
+                    source.setFlowRule(p.flowRule);
+                }
             }
 
             // add missing exceptional edges
-            for (Tuple<ExceptionBlockImpl, Integer, ?> p : missingExceptionalEdges) {
-                Integer index = p.b;
-                TypeMirror cause = (TypeMirror) p.c;
-                ExceptionBlockImpl source = p.a;
+            for (MissingEdge p : missingExceptionalEdges) {
+                Integer index = p.index;
+                TypeMirror cause = p.cause;
+                ExceptionBlockImpl source = (ExceptionBlockImpl) p.source;
                 if (index == null) {
                     // edge to exceptional exit
                     source.addExceptionalSuccessor(exceptionalExitBlock, cause);
@@ -3622,11 +3700,12 @@ public class CFGBuilder {
             addLabelForNextNode(trueStart);
             Node trueExpr = scan(tree.getTrueExpression(), p);
             trueExpr = conditionalExprPromotion(trueExpr, exprType);
-            extendWithExtendedNode(new UnconditionalJump(merge));
+            extendWithExtendedNode(new UnconditionalJump(merge, Store.FlowRule.BOTH_TO_THEN));
 
             addLabelForNextNode(falseStart);
             Node falseExpr = scan(tree.getFalseExpression(), p);
             falseExpr = conditionalExprPromotion(falseExpr, exprType);
+            extendWithExtendedNode(new UnconditionalJump(merge, Store.FlowRule.BOTH_TO_ELSE));
 
             addLabelForNextNode(merge);
             Node node = new TernaryExpressionNode(tree, condition, trueExpr, falseExpr);

--- a/framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java
+++ b/framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java
@@ -677,19 +677,21 @@ public abstract class CFAbstractTransfer<
         return new RegularTransferResult<>(finishValue(value, store), store);
     }
 
-    /** The resulting abstract value is the merge of the 'then' and 'else' branch. */
     @Override
     public TransferResult<V, S> visitTernaryExpression(
             TernaryExpressionNode n, TransferInput<V, S> p) {
         TransferResult<V, S> result = super.visitTernaryExpression(n, p);
-        S store = result.getRegularStore();
+        S thenStore = result.getThenStore();
+        S elseStore = result.getElseStore();
         V thenValue = p.getValueOfSubNode(n.getThenOperand());
         V elseValue = p.getValueOfSubNode(n.getElseOperand());
         V resultValue = null;
         if (thenValue != null && elseValue != null) {
+            // The resulting abstract value is the merge of the 'then' and 'else' branch.
             resultValue = thenValue.leastUpperBound(elseValue);
         }
-        return new RegularTransferResult<>(finishValue(resultValue, store), store);
+        V finishedValue = finishValue(resultValue, thenStore, elseStore);
+        return new ConditionalTransferResult<>(finishedValue, thenStore, elseStore);
     }
 
     /** Reverse the role of the 'thenStore' and 'elseStore'. */

